src/GovUk.Education.ExploreEducationStatistics.Public.Data.Processor/Services/DataSetVersionService.cs

```diff
@@ -123,4 +142,264 @@ private void DeleteParquetFiles(DataSetVersion dataSetVersion)
+                message: ValidationMessages.FileNotFound,
+                releaseFileId: releaseFileId
+            ))
+            : releaseFile;
+    }
+
+    private async Task<Either<ActionResult, Unit>> ValidateReleaseFile(
+        ReleaseFile releaseFile,
+        CancellationToken cancellationToken)
+    {
+        // ReleaseFile must not already have a DataSetVersion
+        if (await publicDataDbContext.DataSetVersions.AnyAsync(
+                dsv => dsv.ReleaseFileId == releaseFile.Id,
+                cancellationToken: cancellationToken))
+        {
+            return ValidationUtils.ValidationResult(
+            [
+                CreateReleaseFileIdError(
+                    message: ValidationMessages.FileHasApiDataSetVersion,
+                    releaseFileId: releaseFile.Id)
+            ]);
+        }
+
+        // ReleaseFile must relate to a ReleaseVersion in Draft approval status
+        if (releaseFile.ReleaseVersion.ApprovalStatus != ReleaseApprovalStatus.Draft)
+        {
+            return ValidationUtils.ValidationResult(
+            [
+                CreateReleaseFileIdError(
+                    message: ValidationMessages.FileReleaseVersionNotDraft,
+                    releaseFileId: releaseFile.Id)
+            ]);
+        }
+
+        List<ErrorViewModel> errors = [];
+
+        // ReleaseFile must relate to a File of type Data
+        if (releaseFile.File.Type != FileType.Data)
+        {
+            errors.Add(CreateReleaseFileIdError(
+                message: ValidationMessages.FileTypeNotData,
+                releaseFileId: releaseFile.Id));
+        }
+
+        // There must be a ReleaseFile related to the same ReleaseVersion and Subject with File of type Metadata
+        if (!await contentDbContext.ReleaseFiles
+                .Where(rf => rf.ReleaseVersionId == releaseFile.ReleaseVersionId)
+                .Where(rf => rf.File.SubjectId == releaseFile.File.SubjectId)
+                .Where(rf => rf.File.Type == FileType.Metadata)
+                .AnyAsync(cancellationToken: cancellationToken))
+        {
+            errors.Add(CreateReleaseFileIdError(
+                message: ValidationMessages.NoMetadataFile,
+                releaseFileId: releaseFile.Id));
+        }
+
+        return errors.Count == 0 ? Unit.Instance : ValidationUtils.ValidationResult(errors);
+    }
+    
+    private async Task<Either<ActionResult, Unit>> ValidateReleaseFileAndDataSet(
+        ReleaseFile releaseFile,
+        DataSet dataSet,
+        CancellationToken cancellationToken)
+    {
+        List<ErrorViewModel> errors = [];
+
+        if (releaseFile.ReleaseVersion.PublicationId != dataSet.PublicationId)
+        {
+            errors.Add(CreateReleaseFileIdError(
+                message: ValidationMessages.NextReleaseFileMustBeForSamePublicationAsDataSet,
+                releaseFileId: releaseFile.Id));
+        }
+
+        var historicReleaseFileIds = dataSet
```

Comment:
We don't typically use 'historical' to describe things, and given my earlier comment around the validation message, perhaps `previousReleaseFileIds` would be more consistent.

Same could be applied to `historicalReleaseIds` (i.e. `previousReleaseIds`) below